[RFC] Swift 4, simplifying response parsing, and no more Argo?

sharplet opened the pull request:
This is an experiment to see what it might be like if Swish drops support for Argo by requiring Swift 4 and making use of the new `JSONDecoder`. It also makes some changes to the way URL responses are handled, removing some optionality by encoding some assumptions described in the `URLSession.dataTask(with:completionHandler:)` documentation. See commits for more details.

sharplet committed:
Update minimum swift version to 4

sharplet committed:
Assume Data and URLResponse are non-nil

The documentation for `URLSession.dataTask(with:completionHandler:)`
states that the `data` and `error` parameters to the completion handler
are mutually exclusive:

> If the request completes successfully, the `data` parameter of the
> completion handler block contains the resource data, and the `error`
> parameter is `nil`. If the request fails, the `data` parameter is
> `nil` and the `error` parameter contain information about the failure.
> If a response from the server is received, regardless of whether the
> request completes successfully or fails, the `response` parameter
> contains that information.

It also follows from this that if the `data` parameter is non-nil, the
`request` parameter will also be non-nil.

This change updates `HTTPResponse` to more accurately model this,
removing some unnecessary `nil` checks, and force-downcasts the
`URLResponse` to `HTTPURLResponse`.

sharplet committed:
Delete deprecations

sharplet committed:
Remove Argo in favour of JSONDecoder

sharplet committed:
Add some missing tests that weren't in it() blocks

sharplet committed:
Use throws instead of Result in Request.parse(_:)

This allows Swish to take full control over how `Result`s are
constructed internally, and gives clients the flexibility to throw
whatever error they like, knowing it will be handled by Swish.

This resolves a crash where unsafe use of `Result`'s with `SwishError`
causes a force-cast to fail.

sharplet committed:
Use an inline struct instead of fileprivate

sharplet commented:
I'm going to make a `swift-4` branch and target this PR there, so we can keep breaking changes isolated from `master` for now.

gfontenot commented on Source/Extensions/Deprecations+Removals.swift:
```diff
@@ -1,24 +0,0 @@
-  public static func fromDecoded(_: Decoded<T>) -> Result<T, SwishError> {
-    fatalError("unavailable")
-  }
-}
```
What do you think about leaving this file in place for future deprecations, but keeping it empty?

sharplet replied:
Sure

gfontenot reviewed (approved):
Sad to see Argo go here, but all in all this makes a lot of sense.

gfontenot committed:
Remove @testable from tests

We aren't accessing any internal properties or methods, so we don't actually
need to use @testable here.

gfontenot committed:
Add local vimrc

This adds some basic config for working with the project from vim with
vim-xcode:

https://github.com/gfontenot/vim-xcode

njdehoog commented:
This looks good. Are there plans to make this part of a new release? It looks like it's been on hold for a while.

klaaspieter commented:
@sharplet I rebased this for you. Okay to merge?

sharplet committed:
Upgrade Xcode version on CI

sharplet committed:
Update Apple TV simulator name when running tests

sharplet commented:
@klaaspieter Thanks! I've fixed up the CI issues as well, so I'm going to go ahead and merge this.

sharplet commented:
I'll look into putting out an official release later this week. In the meantime, this is available on `master`.
